server/src/main/java/org/elasticsearch/search/lookup/SourceFilter.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.lookup;
+
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.io.stream.BytesStreamOutput;
+import org.elasticsearch.common.util.CollectionUtils;
+import org.elasticsearch.common.xcontent.support.XContentMapValues;
+import org.elasticsearch.xcontent.XContent;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xcontent.XContentParserConfiguration;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+
+/**
+ * Implements source filtering based on a list of included and excluded fields.  To use,
+ * construct a SourceFilter and pass it to {@link Source#filter(SourceFilter)}
+ */
+public final class SourceFilter {
+
+    private static final Function<Map<String, Object>, Map<String, Object>> EMPTY_FILTER = v -> v;
```

Comment:
You may find it slightly more readable to have an `empty` version of this class. I see you are using this `EMPTY_FILTER` as a sentinel to encode the "do nothing" behavior. But when I read this class I thought "oh, I should tell Alan that the JVM will generate this constant for him. Maybe he doesn't know." Then I read further down and saw it was a sentinel. Which is fine. I thought "oh, he should leave a comment on this so no one else wonders why it's a constant instead of inline." And then I thought - oh, maybe it'd be more clear if this had an implementation that always did nothing. I dunno if that's actually true, but maybe.